consensus/src/main/java/org/ethereum/beacon/consensus/transition/BlockTransition.java

```diff
@@ -0,0 +1,167 @@
+  private final SpecHelpers specHelpers;
+
+  public BlockTransition(SpecHelpers specHelpers) {
+    this.specHelpers = specHelpers;
+    this.spec = specHelpers.getChainSpec();
+  }
+
+  @Override
+  public BeaconStateEx apply(BeaconBlock block, BeaconStateEx stateEx) {
+    MutableBeaconState state = stateEx.getCanonicalState().createMutableCopy();
+
+    /*
+      RANDAO
+      Let proposer = state.validator_registry[get_beacon_proposer_index(state, state.slot)].
+      Set state.latest_randao_mixes[state.slot % LATEST_RANDAO_MIXES_LENGTH] =
+        hash(xor(state.latest_randao_mixes[state.slot % LATEST_RANDAO_MIXES_LENGTH], block.randao_reveal))
+      Set proposer.randao_commitment = block.randao_reveal.
+      Set proposer.randao_layers = 0.
+    */
+
+    int proposerIdx = specHelpers.get_beacon_proposer_index(state, state.getSlot()).getValue();
+    state.withValidatorRecord(
+        proposerIdx,
+        vb -> vb.withRandaoCommitment(block.getRandaoReveal()).withRandaoLayers(UInt64.ZERO));
+
+    /*
+     Deposit root
+     If block.deposit_root is deposit_root_vote.deposit_root for some deposit_root_vote
+       in state.deposit_root_votes, set deposit_root_vote.vote_count += 1.
+     Otherwise, append to state.deposit_root_votes a
+       new DepositRootVote(deposit_root=block.deposit_root, vote_count=1).
+    */
+
+    int depositIdx = -1;
+    for (int i = 0; i < state.getDepositRootVotes().size(); i++) {
+      if (block.getDepositRoot().equals(state.getDepositRootVotes().get(i).getDepositRoot())) {
+        depositIdx = i;
+        break;
+      }
+    }
+    if (depositIdx >= 0) {
+      state.withDepositRootVote(
+          depositIdx,
+          vote -> new DepositRootVote(vote.getDepositRoot(), vote.getVoteCount().increment()));
+    } else {
+      state.withNewDepositRootVote(new DepositRootVote(block.getDepositRoot(), UInt64.valueOf(1)));
+    }
+
+    /*
+       For each proposer_slashing in block.body.proposer_slashings:
+       Run penalize_validator(state, proposer_slashing.proposer_index).
+    */
+    for (ProposerSlashing proposer_slashing : block.getBody().getProposerSlashings()) {
+      specHelpers.penalize_validator(state, proposer_slashing.getProposerIndex().getValue());
+    }
+
+    /*
+       For each casper_slashing in block.body.casper_slashings:
+       Let slashable_vote_data_1 = casper_slashing.slashable_vote_data_1.
+       Let slashable_vote_data_2 = casper_slashing.slashable_vote_data_2.
+       Let indices(slashable_vote_data) =
+          slashable_vote_data.custody_bit_0_indices + slashable_vote_data.custody_bit_1_indices.
+       Let intersection = [x for x in indices(slashable_vote_data_1) if x in indices(slashable_vote_data_2)].
+       For each validator index i in intersection run penalize_validator(state, i)
+           if state.validator_registry[i].penalized_slot > state.slot.
+    */
+    for (CasperSlashing casper_slashing : block.getBody().getCasperSlashings()) {
```

Comment:
There are `SpecHelpers.indices` and `SpecHelpers.intersection` methods to make this shorter

Reply:
Done